// koishi-utils/dist/string.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function deepen(modifyString) {
    function modifyObject(source) {
        if (typeof source !== 'object' || !source)
            return source;
        if (Array.isArray(source))
            return source.map(modifyObject);
        const result = {};
        for (const key in source) {
            result[modifyString(key)] = modifyObject(source[key]);
        }
        return result;
    }
    return function (source) {
        if (typeof source === 'string') {
            return modifyString(source);
        }
        else {
            return modifyObject(source);
        }
    };
}
exports.camelCase = deepen(source => source.replace(/[_-][a-z]/g, str => str.slice(1).toUpperCase()));
exports.paramCase = deepen(source => source.replace(/_/g, '-').replace(/[A-Z]/g, str => '-' + str.toLowerCase()));
exports.snakeCase = deepen(source => source.replace(/-/g, '_').replace(/[A-Z]/g, str => '_' + str.toLowerCase()));
function capitalize(source) {
    return source.charAt(0).toUpperCase() + source.slice(1);
}
exports.capitalize = capitalize;
