// koishi/dist/utils.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const kleur_1 = __importDefault(require("kleur"));
var logger;
(function (logger) {
    function info(message, logLevel, scope) {
        if (logLevel < 2)
            return;
        console.log(`${kleur_1.default.blue(scope ? 'info:' + scope : 'info')}`, message);
    }
    logger.info = info;
    function error(message, logLevel, scope) {
        if (logLevel < 1)
            return;
        console.log(`${kleur_1.default.red(scope ? 'error:' + scope : 'error')}`, message);
    }
    logger.error = error;
    function warn(message, logLevel, scope) {
        if (logLevel < 2)
            return;
        console.log(`${kleur_1.default.yellow(scope ? 'warning:' + scope : 'warning')}`, message);
    }
    logger.warn = warn;
    function success(message, logLevel, scope) {
        if (logLevel < 1)
            return;
        console.log(`${kleur_1.default.green(scope ? 'success:' + scope : 'success')}`, message);
    }
    logger.success = success;
    function debug(message, logLevel, scope) {
        if (logLevel < 3)
            return;
        console.log(`${kleur_1.default.magenta(scope ? 'debug:' + scope : 'debug')}`, message);
    }
    logger.debug = debug;
})(logger = exports.logger || (exports.logger = {}));
