// koishi-plugin-common/dist/index.js
"use strict";
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
const admin_1 = __importDefault(require("./admin"));
exports.admin = admin_1.default;
const authorize_1 = __importDefault(require("./authorize"));
exports.authorize = authorize_1.default;
const broadcast_1 = __importDefault(require("./broadcast"));
exports.broadcast = broadcast_1.default;
const contextify_1 = __importDefault(require("./contextify"));
exports.contextify = contextify_1.default;
const echo_1 = __importDefault(require("./echo"));
exports.echo = echo_1.default;
const exec_1 = __importDefault(require("./exec"));
exports.exec = exec_1.default;
const exit_1 = __importDefault(require("./exit"));
exports.exit = exit_1.default;
const help_1 = __importDefault(require("./help"));
exports.help = help_1.default;
const info_1 = __importDefault(require("./info"));
exports.info = info_1.default;
const repeater_1 = __importDefault(require("./repeater"));
exports.repeater = repeater_1.default;
const request_handler_1 = __importDefault(require("./request-handler"));
exports.requestHandler = request_handler_1.default;
const respondent_1 = __importDefault(require("./respondent"));
exports.respondent = respondent_1.default;
const welcome_1 = __importDefault(require("./welcome"));
exports.welcome = welcome_1.default;
__export(require("./admin"));
__export(require("./info"));
exports.name = 'common';
function apply(ctx, options = {}) {
    ctx.plugin(request_handler_1.default, options);
    ctx.plugin(repeater_1.default, options.repeater);
    ctx.plugin(respondent_1.default, options.respondent);
    ctx.plugin(welcome_1.default, options.welcomeMessage);
    if (options.echo !== false)
        ctx.plugin(echo_1.default);
    if (options.exec !== false)
        ctx.plugin(exec_1.default);
    if (options.exit !== false)
        ctx.plugin(exit_1.default);
    if (options.help !== false)
        ctx.plugin(help_1.default);
    if (ctx.database) {
        ctx.plugin(authorize_1.default, options);
        if (options.admin !== false)
            ctx.plugin(admin_1.default);
        if (options.contextify !== false)
            ctx.plugin(contextify_1.default);
        if (options.broadcast !== false)
            ctx.plugin(broadcast_1.default, options);
        if (options.info !== false)
            ctx.plugin(info_1.default, options);
    }
}
exports.apply = apply;
